utils/ziputil/zip.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2018 the Service Broker Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ziputil
+
+import (
+	"archive/zip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/GoogleCloudPlatform/gcp-service-broker/utils/stream"
+)
+
+// List writes a ls -la style listing of the zipfile to the given writer.
+func List(z *zip.Reader, w io.Writer) {
+	sw := tabwriter.NewWriter(w, 0, 0, 2, ' ', tabwriter.StripEscape)
+	fmt.Fprintln(sw, "MODE\tSIZE\tNAME")
+	for _, fd := range z.File {
+		fmt.Fprintf(sw, "%s\t%d\t%s\n", fd.Mode().String(), fd.UncompressedSize, fd.Name)
+	}
+	sw.Flush()
+}
+
+// Joins a path for use in a zip file.
+func Join(path ...string) string {
+	return strings.Join(path, "/")
+}
+
+// Clean converts a path into a zip style path by replacing backslashes with
+// forward-slashes and making the paths relative by removing leading "/" and "./".
+func Clean(path ...string) string {
+	joined := filepath.ToSlash(Join(path...))
+	slashStrip := strings.TrimPrefix(joined, "/")
+	dotStrip := strings.TrimPrefix(slashStrip, "./")
+	return dotStrip
+}
+
+// Find returns a pointer to the file at the given path if it exists, otherwise
+// nil.
+func Find(z *zip.Reader, path ...string) *zip.File {
+	name := Join(path...)
+	for _, f := range z.File {
+		if f.Name == name {
+			return f
+		}
+	}
+
+	return nil
+}
+
+// Opens the file at the given path if possible, otherwise returns an error.
+func Open(z *zip.Reader, path ...string) (io.ReadCloser, error) {
+	f := Find(z, path...)
+	if f == nil {
+		fmt.Errorf("no such file: %q", Join(path...))
+	}
+
+	return f.Open()
+}
+
+// Extracts the contents of the zipDirectory to the given OS osDirectory.
+// This routine is overly strict and doesn't allow extracting _any_ files that
+// contain "..".
+func Extract(z *zip.Reader, zipDirectory, osDirectory string) error {
```

Comment:
Not meaning to question the quality of the implementation here, but this all seems like very boiler-plate archive/extract stuff. Thinking towards minimizing the code we have to maintain, I'm curious what the state of prior art is regarding existing libraries in this space?

Reply:
I looked at several libraries trying to get around writing this (mostly so I wouldn't have to write the tests...). They all seem to be deficient in one way or another, not protecting against zip slip, not allowing unarchiving of a sub-directory (only the whole zip), forcing the file extension to be .zip (I opened a bug about that one, but the author won't accept a PR to change it because they think it's a feature), others are GPL, and some don't respect file permissions on extraction.

In an ideal world, I'd really like to use spf13's afero with plugins for zip and gcs: https://github.com/spf13/afero but the project looks dead at this point. Any thoughts about one you like and I might be able to use?

Reply:
I don't have any recommendations, just wanted to verify that all options were thoroughly explored. I trust your judgement here :)